# <span style="color:teal;">CIS 211 Project 7:  SQLite</span>

##### Due 11:00 P.M. Friday Feb 26

This project will give you some experience writing database queries in SQL and connecting to a SQLite database from Python.

### <span style="color:teal">Setup</span>

Download the Sakila database (the file named `sakila211.db`) and move it to the same folder as this IPython notebook.

### <span style="color:teal">Executing Queries in an IPython Notebook</span>

To solve the problems in the first two sections of this project you need to write a SQL query based on the tables in the Sakila211 database.  

To execute a query from an IPython Notebook write the query in quotes as part of a shell command:
```
! sqlite3 sakila211.db "..."
```

For example, to answer the question "What are the names and ratings of the films" your code cell would have
```
! sqlite3 sakila211.db "select title, rating from film"
```

Some of the queries will produce a lot of output.  You can add a `LIMIT` clause to any query if you want to reduce the size of the output.

**Note:** &nbsp; Your grade will not depend on the output format.  If want to have a nicer looking output that includes column headers create a file named `.sqliterc` in your home folder and add these two lines:
```
.header on
.mode column
```

## <span style="color:teal">Part 1:  &nbsp; Queries</span>

The following questions can all be answered using records from a single table.

(1) What are the first and last names of all the customers?

In [1]:
! sqlite3 sakila211.db "select first_name, last_name from customer"

first_name  last_name 
----------  ----------
Mary        Smith     
Patricia    Johnson   
Linda       Williams  
Barbara     Jones     
Elizabeth   Brown     
Jennifer    Davis     
Maria       Miller    
Susan       Wilson    
Margaret    Moore     
Dorothy     Taylor    
Lisa        Anderson  
Nancy       Thomas    
Karen       Jackson   
Betty       White     
Helen       Harris    
Sandra      Martin    
Donna       Thompson  
Carol       Garcia    
Ruth        Martinez  
Sharon      Robinson  
Michelle    Clark     
Laura       Rodriguez 
Sarah       Lewis     
Kimberly    Lee       
Deborah     Walker    
Jessica     Hall      
Shirley     Allen     
Cynthia     Young     
Angela      Hernandez 
Melissa     King      
Brenda      Wright    
Amy         Lopez     
Anna        Hill      
Rebecca     Scott     
Virginia    Green     
Kathleen    Adams     
Pamela      Baker     
Martha      Gonzalez  
Debra       Nelson    
Amanda      Carter    
Stephanie   Mitchell  
Carolyn    

(2) What are the e-mail addresses of customers whose first name is “Willard”?

In [2]:
! sqlite3 sakila211.db "select email from customer where first_name='Willard'"

email                             
----------------------------------
willard.lumpkin@sakilacustomer.org


(3) How many customers are there?

In [3]:
! sqlite3 sakila211.db "select count(*) from customer"

count(*)  
----------
599       


(4) How many customers shop at store number 1?

In [4]:
! sqlite3 sakila211.db "select count(*) from customer where store_id='1'"

count(*)  
----------
326       


(5) How much does it cost to rent the film named “Virtual Spoilers”?

In [5]:
! sqlite3 sakila211.db "select rental_rate from film where title='Virtual Spoilers'"

rental_rate
-----------
4.99       


(6) Do any films have the word “Princess” in the title?

In [6]:
! sqlite3 sakila211.db "select count(*) from film where title like '%Princess%';"

count(*)  
----------
2         


(7) What are the titles of the films that are longer than 180 minutes?

In [7]:
! sqlite3 sakila211.db "select title from film where length>=180"

title          
---------------
Alley Evolution
Analyze Hoosier
Baked Cleopatra
Catch Amistad  
Chicago North  
Confidential In
Conspiracy Spir
Control Anthem 
Crystal Breakin
Darn Forrester 
Frontier Cabin 
Gangs Pride    
Haunting Pianis
Home Pity      
Hotel Happiness
Impact Aladdin 
Intrigue Worst 
Jacket Frisco  
King Evolution 
Lawless Vision 
Love Suicides  
Mixed Doors    
Monsoon Cause  
Moonwalker Fool
Muscle Bright  
Mussolini Spoil
Nash Chocolat  
Pond Seattle   
Records Zorro  
Reds Pocus     
Runaway Tenenba
Saturn Name    
Scalawag Duck  
Searchers Wait 
Smoochy Control
Soldiers Evolut
Something Duck 
Sons Interview 
Sorority Queen 
Star Operation 
Sweet Brotherho
Theory Mermaid 
Wife Turn      
Wild Apollo    
Worst Banger   
Young Language 


(8) How many films have a “G” rating and are less than 60 minutes long?

In [8]:
! sqlite3 sakila211.db "select count(*) from film where rating='G' and length<60"

count(*)  
----------
21        


(9) What is the maximum replacement cost for any film?

In [9]:
! sqlite3 sakila211.db "select replacement_cost from film order by replacement_cost desc limit 1"

replacement_cost
----------------
29.99           


(10) Print a table that lists the different types of ratings and the number of films that have that rating.

In [10]:
! sqlite3 sakila211.db "select rating, count(*) from film group by rating"

rating      count(*)  
----------  ----------
G           178       
NC-17       210       
PG          194       
PG-13       223       
R           195       


## <span style="color:teal">Part 2: &nbsp; Joins</span>

The following queries all require a join of two or more tables. As a hint for how to create the query the table names are shown next to each a question.

(11) What are the film IDs of movies starring Jude Cruise?  &nbsp; &nbsp; [`actor`, `film_actor`]

In [11]:
! sqlite3 sakila211.db "select film_id from film_actor join actor using (actor_id) where first_name='Jude' and last_name='Cruise'"

film_id   
----------
16        
34        
101       
114       
122       
134       
144       
153       
192       
213       
258       
267       
317       
340       
393       
437       
447       
502       
592       
605       
637       
685       
707       
714       
717       
737       
767       
852       
891       
918       


(12) What language was “Ace Goldfinger” filmed in?  &nbsp; &nbsp; [`film`, `language`]

In [12]:
! sqlite3 sakila211.db "select name from language join film using (language_id) where title='Ace Goldfinger'"

name      
----------
Japanese  


(13) How many actors starred in the film named “Scalawag Duck”? &nbsp; &nbsp; [`film`, `film_actor`]

In [13]:
! sqlite3 sakila211.db "select count(*) from film join film_actor using (film_id) where title='Scalawag Duck'"

count(*)  
----------
5         


(14) Produce a table that shows how many films each customer has rented.  &nbsp; &nbsp; [`customer`, `rental`]

In [14]:
! sqlite3 sakila211.db "select first_name, last_name, count(*) from customer join rental using (customer_id) group by first_name, last_name"

first_name  last_name   count(*)  
----------  ----------  ----------
Aaron       Selby       24        
Adam        Gooch       22        
Adrian      Clary       19        
Agnes       Bishop      23        
Alan        Kahn        26        
Albert      Crouse      23        
Alberto     Henning     21        
Alex        Gresham     33        
Alexander   Fennell     36        
Alfred      Casillas    26        
Alfredo     Mcadams     20        
Alice       Stewart     33        
Alicia      Mills       21        
Allan       Cornish     19        
Allen       Butterfiel  21        
Allison     Stanley     27        
Alma        Austin      35        
Alvin       Deloach     29        
Amanda      Carter      27        
Amber       Dixon       27        
Amy         Lopez       29        
Ana         Bradley     34        
Andre       Rapp        28        
Andrea      Henderson   22        
Andrew      Purdy       27        
Andy        Vanhorn     25        
Angel       Barclay 

(15) How many films did the customer named Smith rent?  &nbsp; &nbsp; [`customer`, `rental`]

In [15]:
! sqlite3 sakila211.db "select count(*) from customer join rental using (customer_id) where last_name='Smith'"

count(*)  
----------
32        


(16) Which customers have not returned films (*i.e.* the return_date field in the rental table is null)?  &nbsp; &nbsp; [`customer`, `rental`]

In [16]:
! sqlite3 sakila211.db "select first_name, last_name from customer join rental using (customer_id) where return_date is null"

first_name  last_name 
----------  ----------
Gail        Knight    
Gregory     Mauldin   
Louise      Jenkins   
Willie      Howell    
Emily       Diaz      
Laurie      Lawrence  
Lisa        Anderson  
Freddie     Duggan    
Heather     Morris    
Roland      South     
Natalie     Meyer     
Scott       Shelley   
Louis       Leone     
Cathy       Spencer   
Guy         Brownlee  
Justin      Ngo       
Jerry       Jordon    
Alicia      Mills     
Jeanne      Lawson    
Dustin      Gillette  
Morris      Mccarter  
Judy        Gray      
Julie       Sanchez   
Ramona      Hale      
Sonia       Gregory   
Margie      Wade      
Zachary     Hite      
Jordan      Archuleta 
Andy        Vanhorn   
Elmer       Noe       
Peggy       Myers     
Sergio      Stanfield 
Laura       Rodriguez 
Anna        Hill      
Gail        Knight    
Marilyn     Ross      
Mildred     Bailey    
Grace       Ellis     
Barry       Lovelace  
Edward      Baugh     
Cassandra   Walters   
Bill       

(17) Which actors starred in “Splash Gump”?  &nbsp; &nbsp; [`film`, `film_actor`, `actor`]

In [17]:
! sqlite3 sakila211.db "select first_name, last_name from actor join film_actor using (actor_id) join film using (film_id) where title='Splash Gump'"

first_name  last_name 
----------  ----------
Penelope    Guiness   
Jennifer    Davis     
Uma         Wood      
Cameron     Streep    
Christian   Neeson    
Ray         Johansson 
Gene        Willis    
Morgan      Williams  
Humphrey    Willis    
Al          Garland   


(18) Which films has Penelope Guiness starred in?  &nbsp; &nbsp; [`film`, `film_actor`, `actor`]

In [18]:
! sqlite3 sakila211.db "select title from film join film_actor using (film_id) join actor using (actor_id) where first_name='Penelope' and last_name='Guiness'"

title           
----------------
Academy Dinosaur
Anaconda Confess
Angels Life     
Bulworth Command
Cheaper Clyde   
Color Philadelph
Elephant Trojan 
Gleaming Jawbrea
Human Graffiti  
King Evolution  
Lady Stage      
Language Cowboy 
Mulholland Beast
Oklahoma Jumanji
Rules Human     
Splash Gump     
Vertigo Northwes
Westward Seabisc
Wizard Coldblood


(19) What are the names of the films rented by the customer named Knight?  &nbsp; &nbsp; [`customer`, `rental`, `inventory`, `film`]

In [19]:
! sqlite3 sakila211.db "select title from film join inventory using (film_id) join rental using (inventory_id) join customer using (customer_id) where last_name='Knight'"

title        
-------------
Jericho Mulan
Frost Head   
Clerks Angels
Wife Turn    
Expecations N
Lawless Visio
Soup Wisdom  
Heartbreakers
Bikini Borrow
Attacks Hate 
Cupboard Sinn
Princess Gian
Cyclone Famil
Platoon Insti
Knock Warlock
Excitement Ev
Right Cranes 
Hyde Doctor  
Fatal Haunted
Chasing Fight
Falcon Volume
Tomorrow Hust
Half Outfield
Horror Reign 
Borrowers Bed


## <span style="color:teal">Part 3: &nbsp; Monthly Report</span>

For this part of the project you will write a report generator that will print a summary of monthly activity for a customer.  The report should show a list of all the films rented during the month along with a total amount due.
This example shows the activity for the customer named Leo Ebert for the month of June, 2005.

<pre>
--- Sakila DVD Rentals ---

Monthly Report for Leo Ebert 

Lonely Elephant                 06/16/2005    $2.99
                   ** late fee  06/24/2005    $2.99
Enemy Odds                      06/18/2005    $4.99
                   ** late fee  06/27/2005    $4.99
Song Hedwig                     06/20/2005    $0.99
                   ** late fee  06/26/2005    $0.99
Wash Heavenly                   06/21/2005    $4.99
Hurricane Affair                06/21/2005    $2.99

Monthly Total:  $25.92
</pre>

To compute the total, the report generator needs to add up all the rental fees, and for any film returned late, add an additional rental fee as the late charge. In the example, Leo rented *Lonely Elephant* on June 16 but didn’t return it until June 24. It was a 3-day rental, so he was charged an additional \$2.99 (it was 5 days late, so we could have charged him \$4.98 in late fees, but we’ll keep it simple and just charge the basic fee as the late fee).

We've broken the problem into separate parts so they can be implemented and tested separately.  Your job is to fill in the code cells so each part passes the auto-grader tests.

**Note:** &nbsp; There is no documentation requirement for this project.

### <span style="color:teal">Database Connection</span>

To simplify the code for this program you can define the database connection as a global variable.  The following code cell imports the libraries you need and defines the database connection.

In [4]:
from sqlite3 import connect

db = connect('sakila211.db')

### <span style="color:teal">Dates</span>

Dates in a SQLite database are stored as strings.  The function below uses a Python library named `datetime` to parse a date string and return a string that can be printed as part of the report.  Read the online documentation for the `datetime` module if you want to learn more about date formats and functions that parse and format dates.

**Note:** &nbsp; The format string is a global variable -- you can use this same format in your function that checks to see if a CD is overdue.

In [5]:
from datetime import datetime

date_format = "%Y-%m-%d %H:%M:%S.%f"

def mmddyyyy(date):
    d = datetime.strptime(date, date_format)
    return '{:02d}/{:02d}/{:02d}'.format(d.month, d.day, d.year)

In [6]:
assert mmddyyyy('2005-02-26 20:59:35.000') == '02/26/2005'

### <span style="color:teal">Customer Name</span>

Fill in the definition of the `customer_name` function.  It will be passed a string with a last name, and it should return a list of tuples of the form `(first, last)` with the first and last names of all customers with the same last name.  If there are no customers with that name the list should be empty.

In [7]:
def customer_names(s):
    '''
    Takes an name (s) to return a tuple with the full name of a customer from skalia database
    '''
    fullnames = []
    lastnames = db.execute("select first_name, last_name from customer")
    for name in lastnames:
        if name[1].lower() == s:
            fullnames.append(name)
    return fullnames


##### <span style="color:red">Autograder Tests:</span>

**Do not edit, copy, or delete these cells.**

In [8]:
assert customer_names('ebert') == [('Leo', 'Ebert')]

In [9]:
assert customer_names('smith') == [('Mary', 'Smith')]

In [10]:
assert customer_names('turing') == []

### <span style="color:teal">Film List</span>

Fill in the body of the `film_list` function so it returns a list of all CDs rented by a customer during a specified month. 
The parameters that will be passed to the function are the customer's last name (a string) and the month and year (both integers).  Return a list of tuples that have the following values, in the order shown:
* film title
* rental rate (how much it costs to rent the CD)
* rental date (the date the CD was rented)
* return date (the date the CD was returned)
* rental duration (the number of days the CD can be out before a late fee is charged) 

All of the information you need can be retrieved with a single query that joins the rental, inventory, film, and customer tables. Note that this is the same query used to answer Question 19 above, except with different column names.

**Hint:** &nbsp; The dates in the database are strings.  Think about pattern matching when you write the WHERE clause that specifies the date range.


In [11]:
def film_list(name, month, year): 
    '''
    Takes a customer name, month, and year and returns a list with the movies the given customer has rented and returned, 
    the rate, and the rental duration 
    '''
    month, year = str(month).zfill(2), str(year)
    
    filmlist = []
    
    query = 'select title, rental_rate, rental_date, return_date, rental_duration from film join inventory using (film_id) \
    join rental using (inventory_id) join customer using (customer_id) where last_name=? and strftime("%m", rental_date)=? and \
    strftime("%Y", rental_date)=?'
    
    info = db.execute(query, (name, month, year))
    
    for line in info:
        filmlist.append(line)
    return filmlist

##### <span style="color:red">Autograder Tests:</span>

**Do not edit, copy, or delete these cells.**

In [12]:
leo = film_list('Ebert', 6, 2005)
assert len(leo) == 5
assert leo[0] == (('Lonely Elephant', 2.99, '2005-06-16 20:59:35.000', '2005-06-24 15:31:35.000', 3))

In [13]:
mary = film_list('Smith', 8, 2005)
assert len(mary) == 11
assert mary[0] == ('Adaptation Holes', 2.99, '2005-08-01 08:51:04.000', '2005-08-10 12:12:04.000', 7)

### <span style="color:teal">Overdue CDs</span>

To decide whether a CD was returned late you need to subtract the rental date from the return date and see if the difference is greater than the rental duration. The easiest way to do this is to use Python’s `datetime` library.

Using a call to `strptime` (see the body of the `mmddyyyy` function defined above) create `datetime` objects for the rental date and the return date. Then subtract the rental date from the return date to get a `timedelta` object:
```
    diff = returned - rented
```
Now you can get attributes of the object named `diff` to tell you how far apart the dates were, for example `diff.days` will tell you the time difference in days.

Fill in the body of the `overdue` function.  The parameters are a rental date, a return date, and the rental duration.  Return TRUE if the CD was not returned or if the number of days it was checked out is greater than the rental duration.

In [18]:
def overdue(rent_date, return_date, duration):
    '''
    This takes two dates and a rental duration and returns true if the rental is overdue
    '''
    diff = datetime.strptime(return_date, date_format) - datetime.strptime(rent_date, date_format)
    return diff.days >= duration

In [19]:
assert overdue('2005-06-16 20:59:35.000', '2005-06-19 20:59:35.000', 2)

In [20]:
assert not overdue('2005-06-16 20:59:35.000', '2005-06-17 20:59:35.000', 2)

In [21]:
assert not overdue('2005-06-16 20:59:35.000', '2005-06-19 20:59:34.000', 2)

AssertionError: 

### <span style="color:teal">Report Generator</span>

Fill in the body of the function named `sakila_report` so it prints a monthly report.  The arguments passed to the function will be a customer's last name (a string) and a month and year (both integers).

If the customer is not in the database, or if there are two customers with the same last name, return None.  Otherwise print a list of films and late fees and the total amount due.

You can format the report however you would like, as long as the following items appear somewhere in the output:
* every fee should be printed with a dollar sign and an amount, for example `$2.99` or `25.92`
* the total amound due should be printed on a line that includes the word `"Total"` followed by a dollar amount on the same line

In [34]:
def sakila_report(customer, month, year):
    '''
    This takes a customer name, month and year to create a report of the month on the fees owed by given customer
    '''
    filmlist = film_list(customer, month, year)
    total = 0
    for line in filmlist:
        print('{} ${} {}'.format(line[0], line[1], line[2]))
        total += line[1]
        if overdue(line[2], line[3], line[4]):
            print('** late fee {} ${}'.format(line[3], line[1]))
            total += line[1]
    if total == 0:
        return None
    print('Monthly Total:  ${:.2f}'.format(total))

##### <span style="color:red">Tests</span>

You can use code cells in this section (and add more if you wish) to test your functions.

##### <span style="color:red">Autograder Tests</span>

The tests for the `sakila_report` function will call the function and while using the IPython `capture` command to save everything the function prints.  We'll then use a function that counts the number of times a dollar amount (from either a rental fee or a late fee) occurs and make sure the total amount is the expected value.

**Do not edit, copy, or delete these cells.**

In [35]:
import re

def check_report(text, expected_count, expected_total):
    nums = []
    for line in text.split('\n'):
        has_total = re.search(r'Total.*(\$\d+\.\d\d)', line)
        if has_total:
            assert has_total.group(1) == expected_total
        else:
            for s in re.findall(r'\$\d+\.\d\d', line):
                nums.append(s)
    assert(len(nums) == expected_count)
    total = 0.0
    for s in nums:
        total += float(s[1:])
    assert(round(total,2)) == round(float(expected_total[1:]), 2)
    return True

In [36]:
%%capture res
sakila_report('Ebert', 6, 2005)

In [37]:
assert check_report(res.stdout, 8, '$25.92')

In [38]:
%%capture res
sakila_report('Smith', 8, 2005)

In [39]:
assert check_report(res.stdout, 15, '$32.85')

In [40]:
%%capture res
sakila_report('Turing', 6, 2005)

In [41]:
assert len(res.stdout) == 0